python/ray/dashboard/memory.py

```diff
@@ -0,0 +1,280 @@
+
+
+class SortingType(Enum):
+    PID = 1
+    OBJECT_SIZE = 3
+    REFERENCE_TYPE = 4
+
+
+class GroupByType(Enum):
+    NODE_ADDRESS = 2
+
+
+class ReferenceType:
+    # We don't use enum because enum is not json serializable.
+    ACTOR_HANDLE = "ACTOR_HANDLE"
+    PINNED_IN_MEMORY = "PINNED_IN_MEMORY"
+    LOCAL_REFERENCE = "LOCAL_REFERENCE"
+    USED_BY_PENDING_TASK = "USED_BY_PENDING_TASK"
+    CAPTURED_IN_OBJECT = "CAPTURED_IN_OBJECT"
+    UNKNOWN_STATUS = "UNKNOWN_STATUS"
+
+
+class MemoryTableEntry:
+    def __init__(self, *, object_ref: dict, node_address: str, is_driver: bool,
+                 pid: int):
+        # worker info
+        self.is_driver = is_driver
+        self.pid = pid
+        self.node_address = node_address
+
+        # object info
+        self.object_size = int(object_ref.get("objectSize", -1))
+        self.call_site = object_ref.get("callSite", "<Unknown>")
+        self.object_id = ray.ObjectID(
+            decode_object_id_if_needed(object_ref["objectId"]))
+
+        # reference info
+        self.local_ref_count = int(object_ref.get("localRefCount", 0))
+        self.pinned_in_memory = bool(object_ref.get("pinnedInMemory", False))
+        self.submittedd_task_ref_count = int(
+            object_ref.get("submittedTaskRefCount", 0))
+        self.contained_in_owned = [
+            ray.ObjectID(decode_object_id_if_needed(object_id))
+            for object_id in object_ref.get("containedInOwned", [])
+        ]
+        self.reference_type = self._get_reference_type()
+
+    def is_valid(self) -> bool:
+        if (not self.pinned_in_memory and self.local_ref_count == 0
+                and self.submittedd_task_ref_count == 0
+                and len(self.contained_in_owned) == 0):
+            return False
+        elif self.object_id.is_nil():
+            return False
+        else:
+            return True
+
+    def group_key(self, group_by_type: GroupByType):
+        if group_by_type == GroupByType.NODE_ADDRESS:
+            return self.node_address
+        else:
+            raise ValueError(
+                "group by type {} is invalid.".format(group_by_type))
+
+    def _get_reference_type(self) -> str:
+        if self._is_object_id_actor_handle():
+            return ReferenceType.ACTOR_HANDLE
+        if self.pinned_in_memory:
+            return ReferenceType.PINNED_IN_MEMORY
+        elif self.submittedd_task_ref_count > 0:
+            return ReferenceType.USED_BY_PENDING_TASK
+        elif self.local_ref_count > 0:
+            return ReferenceType.LOCAL_REFERENCE
+        elif len(self.contained_in_owned) > 0:
+            return ReferenceType.CAPTURED_IN_OBJECT
+        else:
+            return ReferenceType.UNKNOWN_STATUS
+
+    def _is_object_id_actor_handle(self) -> bool:
+        object_id_hex = self.object_id.hex()
+        random_bits = object_id_hex[:16]
+        actor_random_bits = object_id_hex[16:24]
+
+        # random (8B) | ActorID(6B) | flag (2B) | index (6B)
```

Comment:
It would be great if we could pull these constants from the ID definition somehow. I don't want to accidentally break the dashboard if I change the ID bit allocations.

Reply:
Fixed!